src/javascript/app_2/App/Components/Elements/tabs.jsx

```diff
@@ -0,0 +1,147 @@
+import classNames from 'classnames';
+import React      from 'react';
+import PropTypes  from 'prop-types';
+
+class Tabs extends React.PureComponent {
+    constructor(props) {
+        super(props);
+        this.state = { active_tab_index: '1' };
+    }
+    setActiveTab(index) {
+        this.setState({active_tab_index: index});
+    };
+    render() {
+        const TabContents = this.props.list[this.state.active_tab_index].content;
+        const tab_container_class = classNames('tab-container', this.props.alignment);
+        const tab_header_class = (icon_name) => classNames({
+            icon: icon_name },
+            icon_name,
+        );
+        return (
+            <div className={tab_container_class}>
+                <TabsWrapper
+                    active={this.state.active_tab_index}
+                    onChange={active => this.setActiveTab(active)}
+                >
+                    {Object.keys(this.props.list).map(key => (
+                        <React.Fragment key={key}>
+                            <span
+                                className={tab_header_class(this.props.list[key].icon)}
+                                title={this.props.list[key].header}
+                            >
+                                {this.props.list[key].header}
+                            </span>
+                        </React.Fragment>
+                    ))}
+                </TabsWrapper>
+                <TabContents />
+            </div>
+        );
+    }
+};
+
+class TabsWrapper extends React.PureComponent {
+    constructor(props) {
+        super(props);
+        this.state = {
+            sizes: {},
+        };
+        this.els = {};
+    }
+
+    componentDidMount() {
+        this.getSizes();
+        window.addEventListener('resize', this.getSizes.bind(this));
+    }
+
+    componentDidUpdate(prevProps) {
+        // component re-renders twice, pending better workaround
+        if (prevProps.children !== this.props.children && prevProps.active !== this.props.active) {
+            this.getSizes();
+        }
+    }
+
+    componentWillUnmount() {
+        window.removeEventListener('resize', this.getSizes.bind(this));
+    }
+
+    getSizes() {
+        const rootBounds = this.root.getBoundingClientRect();
+
+        const sizes = {};
+
+        Object.keys(this.els).forEach((key) => {
+            const el = this.els[key];
+            const bounds = el.getBoundingClientRect();
+
+            const left = bounds.left - rootBounds.left;
+            const right = rootBounds.right - bounds.right;
+
+            sizes[key] = {left, right};
+        });
+
+        this.setState({sizes});
+        return sizes;
```

Comment:
returning sizes seems unnecessary? 🤔

Reply:
yep, forgot to remove it after testing. thanks!